File: C++/Graphs/Find the Celebrity.cpp
/* The knows API is defined for you.
      bool knows(int a, int b); */
//Approach - 1
//TC: O(n) where n is number of persons
//SC: O(1)
class Solution {
    private:
        int i_NumberOfPeople ;
        int i_NumberOfCelebrity ;
        bool b_isCelebrity(int i) ;
public:
    int findCelebrity(int n) {
        i_NumberOfPeople = n ;
        for(int i = 0;i<n;i++)
        {
            if ( knows(i_NumberOfCelebrity,i) )
            {
                i_NumberOfCelebrity = i;
            }
        }
        if ( b_isCelebrity(i_NumberOfCelebrity) )
        {
            return i_NumberOfCelebrity ;
        }
        
        return -1;
    }
};

bool Solution :: b_isCelebrity ( int i)
{
    for ( int j=0;j<i_NumberOfPeople ; j++)
    {
        if ( i == j )
            continue;
        if ( knows(i,j) == true || knows(j,i) == false )
        {
            return false;
        }
    }
    return true;
    
}

File: Python/Graph/Find the Celebrity.py
# The knows API is already defined for you.
# @param a, person a
# @param b, person b
# @return a boolean, whether a knows b
# def knows(a, b):

class Solution(object):
    def findCelebrity(self, n):
        """
        :type n: int
        :rtype: int
        """
        candidate = 0
        for i in range(1, n):
            if knows(candidate, i):  
                candidate = i      
    
        for i in range(n):
            candidate_knows_i = knows(candidate, i) 
            i_knows_candidate = knows(i, candidate) 
            if i != candidate and (candidate_knows_i or not i_knows_candidate):
                return -1
        return candidate
        
